File: in_Java/0419 Battleships in a Board.java
class Solution {
    public int countBattleships(char[][] board) {
        int res = 0;
        int r;
        int c;
        boolean[][] visited = new boolean[board.length][board[0].length];
        for(int i = 0; i < board.length; i++){
            for(int j = 0; j < board[0].length; j++){
                if(board[i][j] == 'X' && !visited[i][j]){
                    res++;
                    r = i;
                    c = j;
                    while(r < board.length && board[r][j] == 'X'){
                        visited[r][j] = true;
                        r++;
                    }
                    while(c < board[0].length && board[i][c] == 'X'){
                        visited[i][c] = true;
                        c++;
                    }
                }
            }
        }
        return res;
    }
}

File: in_Python/0419 Battleships in a Board.py
class Solution(object):
    def countBattleships(self, board):
        """
        :type board: List[List[str]]
        :rtype: int
        """
        def helper(board, i, j):
            while i + 1 < len(board) and board[i+1][j] == 'X':
                board[i+1][j] = 'S'
                i += 1
            while j + 1 < len(board[0]) and board[i][j+1] == 'X':
                board[i][j+1] = 'S'
                j += 1    
        res = 0
        row = len(board)
        col = len(board[0])
        for i in range(row):
            for j in range(col):
                if board[i][j] == 'X':
                    res += 1
                    helper(board, i, j)
        return res